File: cmajor/parsing/Composite.cpp
// =================================
// Copyright (c) 2019 Seppo Laakko
// Distributed under the MIT license
// =================================

#include <cmajor/parsing/Composite.hpp>
#include <cmajor/parsing/Scanner.hpp>
#include <cmajor/parsing/Visitor.hpp>
#include <cmajor/parsing/Exception.hpp>

namespace cmajor { namespace parsing {

UnaryParser::UnaryParser(const std::u32string& name_, Parser* child_, const std::u32string& info_): Parser(name_, info_), child(child_) 
{
    Own(child);
}

OptionalParser::OptionalParser(Parser* child_): UnaryParser(U"optional", child_, child_->Info() + U"?") 
{
}

Match OptionalParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    Span save = scanner.GetSpan();
    Match match = Child()->Parse(scanner, stack, parsingData);
    if (match.Hit())
    {
        return match;
    }
    scanner.SetSpan(save);
    return Match::Empty();
}

void OptionalParser::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    Child()->Accept(visitor);
    visitor.EndVisit(*this);
}

PositiveParser::PositiveParser(Parser* child_): UnaryParser(U"positive", child_, child_->Info() + U"+")
{
}

Match PositiveParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    Match match = Child()->Parse(scanner, stack, parsingData);
    if (match.Hit())
    {
        for (;;)
        {
            Span save = scanner.GetSpan();
            scanner.Skip();
            Match next = Child()->Parse(scanner, stack, parsingData);
            if (next.Hit())
            {
                match.Concatenate(next);
            }
            else
            {
                scanner.SetSpan(save);
                break;
            }
        }
    }
    return match;
}

void PositiveParser::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    Child()->Accept(visitor);
    visitor.EndVisit(*this);
}

KleeneStarParser::KleeneStarParser(Parser* child_): UnaryParser(U"kleene", child_, child_->Info() + U"*")
{
}

Match KleeneStarParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    Match match = Match::Empty();
    bool first = true;
    for (;;)
    {
        Span save = scanner.GetSpan();
        if (first)
        {
            first = false;
        }
        else
        {
            scanner.Skip();
        }
        Match next = Child()->Parse(scanner, stack, parsingData);
        if (next.Hit())
        {
            match.Concatenate(next);
        }
        else
        {
            scanner.SetSpan(save);
            break;
        }
    }
    return match;
}

void KleeneStarParser::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    Child()->Accept(visitor);
    visitor.EndVisit(*this);
}

ExpectationParser::ExpectationParser(Parser* child_): UnaryParser(U"expectation", child_, child_->Info() + U"!") 
{
}

Match ExpectationParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    Span expectationSpan = scanner.GetSpan();
    Match match = Match::Nothing();
    try
    {
        match = Child()->Parse(scanner, stack, parsingData);
    }
    catch (const ExpectationFailure& ex)
    {
        throw ExpectationFailure(Child()->Info() + ex.Info(), ex.FileName(), ex.GetSpan(), ex.Start(), ex.End());
    }
    if (match.Hit())
    {
        return match;
    }
    else
    {
        throw ExpectationFailure(Child()->Info(), scanner.FileName(), expectationSpan, scanner.Start(), scanner.End());
    }
}

void ExpectationParser::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    Child()->Accept(visitor);
    visitor.EndVisit(*this);
}


TokenParser::TokenParser(Parser* child_): UnaryParser(U"token", child_, child_->Info()) 
{
}

Match TokenParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    scanner.BeginToken();
    Match match = Child()->Parse(scanner, stack, parsingData);
    scanner.EndToken();
    return match;
}

void TokenParser::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    Child()->Accept(visitor);
    visitor.EndVisit(*this);
}

GroupingParser::GroupingParser(Parser* child_) : UnaryParser(U"grouping", child_, child_->Info())
{
}

Match GroupingParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    return Child()->Parse(scanner, stack, parsingData);
}

void GroupingParser::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    Child()->Accept(visitor);
    visitor.EndVisit(*this);
}

BinaryParser::BinaryParser(const std::u32string& name_, Parser* left_, Parser* right_, const std::u32string& info_): Parser(name_, info_), left(left_), right(right_) 
{
    Own(left);
    Own(right);
}

SequenceParser::SequenceParser(Parser* left_, Parser* right_): BinaryParser(U"sequence", left_, right_, left_->Info()) 
{
}

Match SequenceParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    Match leftMatch = Left()->Parse(scanner, stack, parsingData);
    if (leftMatch.Hit())
    {
        scanner.Skip();
        Match rightMatch = Right()->Parse(scanner, stack, parsingData);
        if (rightMatch.Hit())
        {
            leftMatch.Concatenate(rightMatch);
            return leftMatch;
        }
    }
    return Match::Nothing();
}

void SequenceParser::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    Left()->Accept(visitor);
    visitor.Visit(*this);
    Right()->Accept(visitor);
    visitor.EndVisit(*this);
}

AlternativeParser::AlternativeParser(Parser* left_, Parser* right_): BinaryParser(U"alternative", left_, right_, left_->Info() + U" | " + right_->Info()) 
{
}

Match AlternativeParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    Span save = scanner.GetSpan();
    Match leftMatch = Left()->Parse(scanner, stack, parsingData);
    if (leftMatch.Hit())
    {
        return leftMatch;
    }
    scanner.SetSpan(save);
    return Right()->Parse(scanner, stack, parsingData);
}

void AlternativeParser::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    Left()->Accept(visitor);
    visitor.Visit(*this);
    Right()->Accept(visitor);
    visitor.EndVisit(*this);
}

DifferenceParser::DifferenceParser(Parser* left_, Parser* right_): BinaryParser(U"difference", left_, right_, left_->Info() + U" - " + right_->Info()) 
{
}

Match DifferenceParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    Span save = scanner.GetSpan();
    Match leftMatch = Left()->Parse(scanner, stack, parsingData);
    if (leftMatch.Hit())
    {
        Span tmp = scanner.GetSpan();
        scanner.SetSpan(save);
        save = tmp;
        Match rightMatch = Right()->Parse(scanner, stack, parsingData);
        if (!rightMatch.Hit() || rightMatch.Length() < leftMatch.Length())
        {
            scanner.SetSpan(save);
            return leftMatch;
        }
    }
    return Match::Nothing();
}

void DifferenceParser::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    Left()->Accept(visitor);
    visitor.Visit(*this);
    Right()->Accept(visitor);
    visitor.EndVisit(*this);
}

ExclusiveOrParser::ExclusiveOrParser(Parser* left_, Parser* right_): BinaryParser(U"exclusiveOr", left_, right_, left_->Info() + U" ^ " + right_->Info()) 
{
}

Match ExclusiveOrParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    Span save = scanner.GetSpan();
    Match leftMatch = Left()->Parse(scanner, stack, parsingData);
    Span temp = scanner.GetSpan();
    scanner.SetSpan(save);
    save = temp;
    Match rightMatch = Right()->Parse(scanner, stack, parsingData);
    bool match = leftMatch.Hit() ? !rightMatch.Hit() : rightMatch.Hit();
    if (match)
    {
        if (leftMatch.Hit())
        {
            scanner.SetSpan(save);
        }
        return leftMatch.Hit() ? leftMatch : rightMatch;
    }
    return Match::Nothing();
}

void ExclusiveOrParser::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    Left()->Accept(visitor);
    visitor.Visit(*this);
    Right()->Accept(visitor);
    visitor.EndVisit(*this);
}

IntersectionParser::IntersectionParser(Parser* left_, Parser* right_): BinaryParser(U"intersection", left_, right_, left_->Info() + U" & " + right_->Info()) 
{
}

Match IntersectionParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    Span save = scanner.GetSpan();
    Match leftMatch = Left()->Parse(scanner, stack, parsingData);
    if (leftMatch.Hit())
    {
        scanner.SetSpan(save);
        Match rightMatch = Right()->Parse(scanner, stack, parsingData);
        if (leftMatch.Length() == rightMatch.Length())
        {
            return leftMatch;
        }
    }
    return Match::Nothing();
}

void IntersectionParser::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    Left()->Accept(visitor);
    visitor.Visit(*this);
    Right()->Accept(visitor);
    visitor.EndVisit(*this);
}

ListParser::ListParser(Parser* left_, Parser* right_): 
    UnaryParser(U"list", new SequenceParser(left_, new KleeneStarParser(new SequenceParser(right_, left_))), 
        left_->Info() + U" % " + right_->Info()), left(left_), right(right_)
{
}

Match ListParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    return Child()->Parse(scanner, stack, parsingData);
}

void ListParser::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    left->Accept(visitor);
    visitor.Visit(*this);
    right->Accept(visitor);
    visitor.EndVisit(*this);
}

} } // namespace cmajor::parsing

File: devenv/soul.parsing/Composite.cs
/*========================================================================
    Copyright (c) 2011-2012 Seppo Laakko
    http://sourceforge.net/projects/soulparsing/
 
    Distributed under the GNU General Public License, version 3 (GPLv3).
    (See accompanying LICENSE.txt or http://www.gnu.org/licenses/gpl.html)

========================================================================*/

using System;
using System.Collections.Generic;
using System.Text;

namespace soul.parsing
{
    public abstract class UnaryParser : Parser 
    {
        public UnaryParser(Parser child, string info)
            : base(info)
        {
            this.child = child;
        }
        public Parser Child
        {
            get { return child; }
        }
        private Parser child;
    }

    public class OptionalParser : UnaryParser
    {
        public OptionalParser(Parser child)
            : base(child, child.Info + "?")
        {
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            Position save = scanner.Position;
            Match match = Child.Parse(scanner, stack);
            if (match.Hit)
            {
                return match;
            }
            scanner.Position = save;
            return Match.Empty();
        }
        public override void Accept(Visitor visitor)
        {
            visitor.BeginVisit(this);
            Child.Accept(visitor);
            visitor.EndVisit(this);
        }
    }

    public class PositiveParser : UnaryParser
    {
        public PositiveParser(Parser child)
            : base(child, child.Info)
        {
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            Match match = Child.Parse(scanner, stack);
            if (match.Hit)
            {
                for (; ; )
                {
                    Position save = scanner.Position;
                    scanner.Skip();
                    Match next = Child.Parse(scanner, stack);
                    if (next.Hit)
                    {
                        match.Concatenate(next);
                    }
                    else
                    {
                        scanner.Position = save;
                        break;
                    }
                }
            }
            return match;
        }
        public override void Accept(Visitor visitor)
        {
            visitor.BeginVisit(this);
            Child.Accept(visitor);
            visitor.EndVisit(this);
        }
    }

    public class KleeneStarParser : UnaryParser
    {
        public KleeneStarParser(Parser child)
            : base(child, child.Info)
        {
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            Match match = Match.Empty();
            bool first = true;
            for (; ; )
            {
                Position save = scanner.Position;
                if (first)
                {
                    first = false;
                }
                else
                {
                    scanner.Skip();
                }
                Match next = Child.Parse(scanner, stack);
                if (next.Hit)
                {
                    match.Concatenate(next);
                }
                else
                {
                    scanner.Position = save;
                    return match;
                }
            }
        }
        public override void  Accept(Visitor visitor)
        {
            visitor.BeginVisit(this);
            Child.Accept(visitor);
            visitor.EndVisit(this);
        }
    }

    public class ExpectationParser : UnaryParser
    {
        public ExpectationParser(Parser child)
            : base(child, child.Info)
        {
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            Position position = scanner.Position;
            Match match = Match.Nothing();
            try
            {
                match = Child.Parse(scanner, stack);
            }
            catch (ExpectationFailure ex)
            {
                throw new ExpectationFailure(Child.Info + ex.Info, ex.Content, ex.Position, ex.FileName);
            }
            if (match.Hit)
            {
                return match;
            }
            else
            {
                throw new ExpectationFailure(Child.Info, scanner.Content, position, scanner.FileName);
            }
        }
        public override void Accept(Visitor visitor)
        {
            visitor.BeginVisit(this);
            Child.Accept(visitor);
            visitor.EndVisit(this);
        }
    }

    public class TokenParser : UnaryParser
    {
        public TokenParser(Parser child)
            : base(child, child.Info)
        {
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            scanner.BeginToken();
            Match match = Child.Parse(scanner, stack);
            scanner.EndToken();
            return match;
        }
        public override void Accept(Visitor visitor)
        {
            visitor.BeginVisit(this);
            Child.Accept(visitor);
            visitor.EndVisit(this);
        }
    }

    public abstract class BinaryParser : Parser
    {
        public BinaryParser(Parser left, Parser right, string info)
            : base(info)
        {
            this.left = left;
            this.right = right;
        }
        public Parser Left
        {
            get { return left; }
        }
        public Parser Right
        {
            get { return right; }
        }
        private Parser left;
        private Parser right;
    }

    public class SequenceParser : BinaryParser
    {
        public SequenceParser(Parser left, Parser right)
            : base(left, right, left.Info)
        {
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            Match left = Left.Parse(scanner, stack);
            if (left.Hit)
            {
                scanner.Skip();
                Match right = Right.Parse(scanner, stack);
                if (right.Hit)
                {
                    left.Concatenate(right);
                    return left;
                }
            }
            return Match.Nothing();
        }
        public override void Accept(Visitor visitor)
        {
            visitor.BeginVisit(this);
            Left.Accept(visitor);
            visitor.Visit(this);
            Right.Accept(visitor);
            visitor.EndVisit(this);
        }
    }

    public class AlternativeParser : BinaryParser
    {
        public AlternativeParser(Parser left, Parser right)
            : base(left, right, left.Info + " | " + right.Info)
        {
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            Position save = scanner.Position;
            Match left = Left.Parse(scanner, stack);
            if (left.Hit)
            {
                return left;
            }
            scanner.Position = save;
            return Right.Parse(scanner, stack);
        }
        public override void Accept(Visitor visitor)
        {
            visitor.BeginVisit(this);
            Left.Accept(visitor);
            visitor.Visit(this);
            Right.Accept(visitor);
            visitor.EndVisit(this);
        }
    }

    public class DifferenceParser : BinaryParser
    {
        public DifferenceParser(Parser left, Parser right)
            : base(left, right, left.Info + " - " + right.Info)
        {
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            Position save = scanner.Position;
            Match left = Left.Parse(scanner, stack);
            if (left.Hit)
            {
                Position tmp = scanner.Position;
                scanner.Position = save;
                save = tmp;
                Match right = Right.Parse(scanner, stack);
                if (!right.Hit || right.Length < left.Length)
                {
                    scanner.Position = save;
                    return left;
                }
            }
            return Match.Nothing();
        }
        public override void Accept(Visitor visitor)
        {
            visitor.BeginVisit(this);
            Left.Accept(visitor);
            visitor.Visit(this);
            Right.Accept(visitor);
            visitor.EndVisit(this);
        }
    }

    public class ExclusiveOrParser : BinaryParser
    {
        public ExclusiveOrParser(Parser left, Parser right)
            : base(left, right, left.Info + " ^ " + right.Info)
        {
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            Position save = scanner.Position;
            Match left = Left.Parse(scanner, stack);
            Position temp = scanner.Position;
            scanner.Position = save;
            save = temp;
            Match right = Right.Parse(scanner, stack);
            bool match = left.Hit ? !right.Hit : right.Hit;
            if (match)
            {
                if (left.Hit)
                {
                    scanner.Position = save;
                }
                return left.Hit ? left : right;
            }
            return Match.Nothing();
        }
        public override void Accept(Visitor visitor)
        {
            visitor.BeginVisit(this);
            Left.Accept(visitor);
            visitor.Visit(this);
            Right.Accept(visitor);
            visitor.EndVisit(this);
        }
    }

    public class IntersectionParser : BinaryParser
    {
        public IntersectionParser(Parser left, Parser right)
            : base(left, right, left.Info + " & " + right.Info)
        {
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            Position save = scanner.Position;
            Match left = Left.Parse(scanner, stack);
            if (left.Hit)
            {
                scanner.Position = save;
                Match right = Right.Parse(scanner, stack);
                if (left.Length == right.Length)
                {
                    return left;
                }
            }
            return Match.Nothing();
        }
        public override void Accept(Visitor visitor)
        {
            visitor.BeginVisit(this);
            Left.Accept(visitor);
            visitor.Visit(this);
            Right.Accept(visitor);
            visitor.EndVisit(this);
        }
    }

    public class ListParser : UnaryParser
    {
        public ListParser(Parser left, Parser right)
            : base(new SequenceParser(left, new KleeneStarParser(new SequenceParser(right, left))), left.Info + " % " + right.Info)
        {
            this.left = left;
            this.right = right;
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            return Child.Parse(scanner, stack);
        }
        public override void Accept(Visitor visitor)
        {
            visitor.BeginVisit(this);
            left.Accept(visitor);
            visitor.Visit(this);
            right.Accept(visitor);
            visitor.EndVisit(this);
        }
        public Parser Left
        {
            get { return left; }
        }
        public Parser Right
        {
            get { return right; }
        }
        private Parser left;
        private Parser right;
    }
}
